torchao/prototype/sparsity/PartialLinear/partial_linear.py

```diff
@@ -0,0 +1,183 @@
+        >>> output = m(input)
+        >>> print(output.size())
+        torch.Size([128, 30])
+    """
+    __constants__ = ["in_features", "out_features", "top_k", "update_mask_every", "is_sparse_forward"]
+    in_features: int
+    out_features: int
+    top_k: int
+    update_mask_every: int
+    is_sparse_forward: bool
+    weight: Tensor
+
+    def __init__(
+        self,
+        in_features: int,
+        out_features: int,
+        top_k: int = None,
+        bias: bool = True,
+        update_mask_every: int = 50,
+        is_sparse_forward: bool = False,
+        device=None,
+        dtype=None,
+    ) -> None:
+        factory_kwargs = {"device": device, "dtype": dtype}
+        super().__init__()
+
+        self.in_features = in_features
+        self.out_features = out_features
+
+        # Default to half of the input features if not specified
+        if top_k is None:
+            top_k = max(1, in_features // 2)
+
+        if top_k <= 0 or top_k > in_features:
+            raise ValueError(f"top_k must be between 1 and {in_features}, got {top_k}")
+
+        self.top_k = top_k
+        self.update_mask_every = update_mask_every
+        self.is_sparse_forward = is_sparse_forward
+        self._forward_counter = 0
+
+        # Create a full weight matrix
+        self.weight = Parameter(
+            torch.empty((out_features, in_features), **factory_kwargs)
+        )
+
+        # Create a binary mask for the weights
+        mask = torch.ones((out_features, in_features), **factory_kwargs)
+        self.register_buffer('mask', mask)
+
+        if bias:
+            self.bias = Parameter(torch.empty(out_features, **factory_kwargs))
+        else:
+            self.register_parameter("bias", None)
+
+        self.reset_parameters()
+
+    def reset_parameters(self) -> None:
+        # Standard initialization as in nn.Linear
+        init.kaiming_uniform_(self.weight, a=math.sqrt(5))
+
+        if self.bias is not None:
+            fan_in, _ = init._calculate_fan_in_and_fan_out(self.weight)
+            bound = 1 / math.sqrt(fan_in) if fan_in > 0 else 0
+            init.uniform_(self.bias, -bound, bound)
+
+        # Reset the mask to all ones
+        self.mask.fill_(1.0)
+
+        # Initialize the mask to keep only top-k weights
+        self._update_mask()
+
+    def _update_mask(self):
+        with torch.no_grad():
+            # Compute the magnitude of weights
+            weight_mag = self.weight.abs()
+
+            # Create a new binary mask
+            new_mask = torch.zeros_like(self.mask)
+
+            # For each output feature, find the top-k input connections
+            _, top_k_indices = weight_mag.topk(self.top_k, dim=1)
+
+            # Set mask to 1 for top-k weights for each output
+            for i in range(self.out_features):
+                new_mask[i, top_k_indices[i]] = 1.0
+
+            # Update the mask
+            self.mask.copy_(new_mask)
+
+    def _sparse_forward(self, input):
```

Comment:
Why do we need this / what is this doing? I don't understand when you would use this `is_sparse_forward` flag.